community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -142,7 +143,7 @@ TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKerne
     private TransactionHooks.TransactionHooksState hooksState;
     private final KernelStatement currentStatement;
     private final StorageStatement storageStatement;
-    private CloseListener closeListener;
+    private final List<CloseListener> closeListeners = new ArrayList<>( 2 );
```

Comment:
As I've said previously, I'm against these listeners because they make code really hard to understand and debug. Especially when there are multiple of them. `SessionStateMachine.State` operates on `KernelTransaction` which is an internal interface. We can change it's `void #close()` to `long #commit()` that returns transaction id. Then it'll be possible to call `commit()` in the state machine and update `TransactionIdTracker` with the returned tx id.